### ISSUE TITLE
Ludown commands break when ending with a trailing slash

### ISSUE THREAD

**Issue by hansmbakker** [microsoft/botbuilder-tools#948]
## Tool
Name: Ludown
Version: 1.1.0
OS: Windows 10 Build 17763
Environment: Node.js v10.15.1 in Powershell. The bug does not appear in CMD

## Describe the bug
Cannot specify output folder. Running the repro command below makes the tool crash and returns:
```
Output folder C:\Users\MYUSERNAME\Source\Repos\PROJECTFOLDER\DeploymentScripts\nl" --out general.luis does not exist
Stopping further processing.
```
(I removed my username and project folder; they do **not** contain spaces, the project folder contains dots like `CompanyName.ProjectName`)

## To Reproduce
1. Run `ludown parse ToLuis --in ".\CognitiveModels\LUIS\nl\general.lu" --out_folder ".\DeploymentScripts\nl\" --out "general.luis"`

## Expected behavior
Ludown generates `general.luis` file in the specified output folder

## Workaround
When I omit the `--out_folder` parameter, it generates the correct file. I can then manually move the file to the correct folder.

[bug]

**Comment by Zerryth** [comment]
@hansmbakker I'm not able to reproduce the error you got.

However I did notice in my tests that when using the `--out` option, in order to convert it to a `.json` file, I **needed to include `.json` at the end of the name**.
* so if you did `--out general.luis`, it would actually create the contents inside as it would in a `JSON` file, but it would be of type `.luis` file, which is not what we want
___
## Commands ran
I had tried running the command with both `--out_folder` and `--out` together and with the two options separately, and it worked in all combinations:

* `ludown parse toluis --in 9.lu --out_folder myFolder --out test.luis.json`
  * creates `myFolder/test.luis.json`

* `ludown parse toluis --in 9.lu --out_folder myFolder`
  * creates `myFolder/9.json`

* `ludown parse toluis --in 9.lu --out test2.luis.json`
  * creates `myFolder/test2.luis.json`
___
## Additional Info

node: v10.14.1
terminal: Command Prompt, then also tried Windows PowerShell
ludown: 1.1.0
OS: Windows 10 Enterprise

**Comment by Zerryth** [comment]
The `ludown` file I used can be found in the examples in the botbuilder-tools repo [here](https://github.com/Microsoft/botbuilder-tools/tree/master/packages/Ludown/examples).

**Comment by hansmbakker** [comment]
- I did `.luis` instead of `.json` to follow the structure of the Enterprise Bot Template. The content is still json if the filename ends with `.luis`. **It is just a naming convention, nothing functionally different.**
- did you specify `"myfolder"` or `".\myfolder"` or `".\myfolder\mysubfolder"`? I used `".\myfolder\mysubfolder"`.

**Comment by Zerryth** [comment]
I used `myFolder` without `.\`

Tested in Enterprise Bot Template:

_Command Ran_ 
`ludown parse toluis --in CognitiveModels/LUIS/en/general.lu --out_folder DeploymentScripts/myFolder --out anotherTest.withdot.json`

![image](https://user-images.githubusercontent.com/35248895/52306328-7d6c1780-294c-11e9-9503-f98abc55f56b.png)

**Comment by Zerryth** [comment]
Used both **powershell** and **cmd**:  **both succeeded**

However I just tried in **bash** and finally **it failed**

_Command Ran in Bash_
`ludown parse toluis --in CognitiveModels/LUIS/en/general.lu --out_folder DeploymentScripts/myFolder --out test3.json`

Result:
`Sorry unable to open [CognitiveModels/LUIS/en/general.lu]
Stopping further processing.`

**Comment by hansmbakker** [comment]
Also note that I used backslashes (`\`) instead of slashes (`/`) in my path

**Comment by Zerryth** [comment]
* the use of either forward or back slashes worked in powershell (`5.1.17763.134`) and cmd (`Microsoft Windows Version 10.0.17763.253`)
* it appears it was already discovered that the botbuilder-tools coverage is limited in some scenarios, however it was determined its current coverage of OS's and shells is sufficient: #700

**Comment by Zerryth** [comment]
Still unable to repro your error in cmd. 

I tried running the command all the way from `C:` and also created a folder with a `.` in its name, just in case length or punctuation would affect the `ludown parse` command.

_Command Ran_
`ludown parse toluis --in Users\v-asho\source\repos\EnterpriseBot2\EnterpriseBot2\CognitiveModels\LUIS\en\general.lu --out_folder Users\v-asho\source\repos\EnterpriseBot2\EnterpriseBot2\DeploymentScripts\myfolder.withdotname --out longNameDotFolder.json` 
  => creates `longNameDotFolder.json` successfully

and 

`ludown parse toluis --in Users\v-asho\source\repos\EnterpriseBot2\EnterpriseBot2\CognitiveModels\LUIS\en\general.lu --out_folder Users\v-asho\source\repos\EnterpriseBot2\EnterpriseBot2\DeploymentScripts\myfolder.withdotname --out longNameDotFolder2.luis` 
  => creates `longNameDotFolder2.luis` successfully

We could suggest that you ensure that your version node is updated and that you're using either cmd or powershell to run the command (you can see versions I used in comments above)

However without being able to repro the error you get, we can't drill down to a specific fix.
LMK if you think any additional information would be helpful, however

**Comment by hansmbakker** [comment]
My issue with ludown was in powershell (I can still try whether it occurs in cmd), and specifically occurs with the output folder parameter. The issue you’re linking does not seem related to me.

**Comment by hansmbakker** [comment]
Ok, I’ll try to make the reproduction scenario more precise

**Comment by hansmbakker** [comment]
Both on node version v10.10.0 and v10.15.1 the bug appeared.
The bug only appears in PowerShell for me, cmd works fine.

**Comment by Zerryth** [comment]
OK!!!! So, after some messing around with commands in powershell vs. cmd, figured out something:
### ludown breaks when you end a folder with a trailing slash

___
So your original command was this:
 `ludown parse ToLuis --in ".\CognitiveModels\LUIS\nl\general.lu" --out_folder ".\DeploymentScripts\nl\" --out "general.luis"`

It broke with the `--out_folder` because of the trailing `\` after your `nl` folder.

Looking further, you might've typed out the same keys in cmd vs. powershell, however when you tab to auto-complete a folder name, in cmd it does not leave a trailing `\`, but in powershell it does!
Thus is my guess as to why it was working in cmd for you, but not in powershell.

Whereas with me in my testing previously, I typed out the ludown command in cmd first, then copied it over (without trailing slash) to powershell, and it "worked on my machine" just fine.
__
So in summary: don't use trailing slashes with folders when executing ludown commands

Checking with team if this is a bug that we should fix or if they want to go the route of specifying in documentation to just avoid trailing slashes

**Comment by hansmbakker** [comment]
Well done! As the saying goes "the devil is in the details" --> in this case in a trailing backslash...

**Comment by Zerryth** [comment]
As per Steven Gum, labeling to indicate bug and ludown-cli.
Removed Support label. 
___
@cahann Handing over to LUIS team to set priority of when to fix bug

**Summary**: trailing slashes break ludown commands

**Comment by vishwacsena** [comment]
The core issue here is that the commander CLI framework library does not work well with powershell. 

Unfortunately the workaround for now is to not include trailing slashes or use cmd.exe or use cmd.exe within powershell. 

We will continue to explore the possibility of moving to a different CLI framework in the future.
